### PR TITLE
XTypes: Data Representation Refactor

### PR DESCRIPTION
**Pull request by ClaytonCalabrese**
Currently assumes empty annotation is all data representations are valid and empty qos is XCDR1.  This may change depending on our design choices.  This PR should also fix a multitude of tests.

### CODE REVIEW

**Comment by ClaytonCalabrese** [comment]
I still need to make some changes to the Data Rep test based on these changes.

File: dds/DCPS/TopicImpl.h [hunk 1]
```diff
@@ -87,9 +87,15 @@ class OpenDDS_Dcps_Export TopicImpl
 
   /**
    * If the representation sequence of a reader or writer is empty then fill it
-   * either from the QoS of the topic or what the type allows.
+   * from the QoS of the topic.
    */
```
Comment (iguessthislldo) [inline dds/DCPS/TopicImpl.h:91]: The QoS should not be inherited like this. It turned out that was the users job, like it is with the rest of the QoS. I don't think I said that explicitly in my proposed changes though.

Comment (mitza-oci) [reply dds/DCPS/TopicImpl.h:91]: Does it need to happen in QoS_Helper::copy_from_topic_qos (or is that done already)?

Comment (iguessthislldo) [reply dds/DCPS/TopicImpl.h:91]: Yes, but I don't remember if it's already doing it or not.

File: tests/DCPS/DataRepresentation/DataRepresentation.cpp [hunk 2]
```diff
@@ -338,13 +326,18 @@ class Test {
...
     topic_name_ = std::string(type_name) + " Topic";
     if (reprs.size()) {
       topic_name_ += " " + reprs_to_string(reprs);
     }
     topic_ = create_topic(participant_, topic_name_.c_str(), type_name, reprs);
+    if (!topic_ == !expected) {
+        add_result(true);
+    } else {
+      add_result(false);
+    }
```
[the remark this answers is not in the document]
Comment (ClaytonCalabrese) [reply tests/DCPS/DataRepresentation/DataRepresentation.cpp:340]: I think it is a result. If the topic does not have a data representation that works with the type, the topic would not be created.

File: dds/DCPS/TopicImpl.cpp [hunk 3]
```diff
@@ -149,11 +137,9 @@ TopicImpl::enable()
...
   }
 
-  // Make sure Data Representation QoS is initialized
-  DDS::DataRepresentationIdSeq type_allowed_reprs;
-  type_support_->representations_allowed_by_type(type_allowed_reprs);
-  DCPS::check_data_representation_qos(
-    qos_.representation.value, type_allowed_reprs);
+  if (!check_data_representation(qos_.representation.value, false)) {
+    return DDS::RETCODE_PRECONDITION_NOT_MET;
+  }
```
[the remark this answers is not in the document]
Comment (ClaytonCalabrese) [reply dds/DCPS/TopicImpl.cpp:142]: If you enable a topic with qos that is incorrect, shouldn't the enable fail?

File: dds/DCPS/TopicImpl.h (hunk 1 above)
Comment (ClaytonCalabrese) [reply dds/DCPS/TopicImpl.h:91]: It is already there, yes. 

I'll remove the inherit function.

File: dds/DCPS/TopicImpl.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (iguessthislldo) [reply dds/DCPS/TopicImpl.cpp:142]: Well the Topic doesn't use the data representation QoS itself, so I think it's debatable. We're not doing any checks on other QoS here, so I think we should keep in line with that. If there were other QoS that's only used by the readers and writers being checked, then I think we would have to do the same.

[the remark this answers is not in the document]
Comment (ClaytonCalabrese) [reply dds/DCPS/TopicImpl.cpp:142]: You're right. I'll remove that and adjust the DR Test to reflect it.

File: dds/DCPS/DCPS_Utils.cpp [hunk 4]
```diff
@@ -315,6 +315,28 @@ compatibleQOS(const DDS::DataWriterQos * writerQos,
     const CORBA::ULong reader_count = readerQos->representation.value.length();
     const CORBA::ULong writer_count = writerQos->representation.value.length();
     bool found = false;
+
+    if (writer_count == 0 && reader_count == 0){
```
Comment (mitza-oci) [inline dds/DCPS/DCPS_Utils.cpp:319]:
```suggestion
    if (writer_count == 0 && reader_count == 0) {
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 5]
```diff
@@ -1169,8 +1167,11 @@ DataReaderImpl::enable()
     dp_id_ = participant->get_id();
   }
 
-  // Make sure Data Representation QoS is initialized
-  topic_servant_->check_data_representation_qos(qos_.representation.value);
+  if(topic_servant_) {
```
Comment (mitza-oci) [inline dds/DCPS/DataReaderImpl.cpp:1170]:
```suggestion
  if (topic_servant_) {
```

File: dds/DCPS/TopicImpl.cpp [hunk 6]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
   notify_status_condition();
 }
 
-void TopicImpl::check_data_representation_qos(
-  DDS::DataRepresentationIdSeq& qos) const
-{
-  DCPS::check_data_representation_qos(qos, qos_.representation.value);
+bool TopicImpl::check_data_representation(const DDS::DataRepresentationIdSeq& qos_ids, bool is_data_writer) {
+  DDS::DataRepresentationIdSeq type_allowed_reprs;
+  if(!type_support_){
```
Comment (mitza-oci) [inline dds/DCPS/TopicImpl.cpp:222]:
```suggestion
  if (!type_support_){
```

File: dds/DCPS/TopicImpl.cpp [hunk 7]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
+    return true;
+  }
+  type_support_->representations_allowed_by_type(type_allowed_reprs);
+  //default for blank annotation is to allow all types of data representation
+  if(type_allowed_reprs.length() == 0){
+    return true;
+  }
+  //Data Writer will only use the 1st QoS declared
+  if(is_data_writer) {
```
Comment (mitza-oci) [inline dds/DCPS/TopicImpl.cpp:231]:
```suggestion
  if (is_data_writer) {
```

File: dds/DCPS/TopicImpl.cpp [hunk 8]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
-{
-  DCPS::check_data_representation_qos(qos, qos_.representation.value);
+bool TopicImpl::check_data_representation(const DDS::DataRepresentationIdSeq& qos_ids, bool is_data_writer) {
+  DDS::DataRepresentationIdSeq type_allowed_reprs;
+  if(!type_support_){
+    return true;
+  }
+  type_support_->representations_allowed_by_type(type_allowed_reprs);
+  //default for blank annotation is to allow all types of data representation
+  if(type_allowed_reprs.length() == 0){
```
Comment (mitza-oci) [inline dds/DCPS/TopicImpl.cpp:227]:
```suggestion
  if (type_allowed_reprs.length() == 0) {
```

File: dds/DCPS/TopicImpl.cpp [hunk 9]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
+    return true;
+  }
+  //Data Writer will only use the 1st QoS declared
+  if(is_data_writer) {
+    DDS::DataRepresentationId_t id=DDS::XCDR_DATA_REPRESENTATION;
```
Comment (mitza-oci) [inline dds/DCPS/TopicImpl.cpp:232]:
```suggestion
    DDS::DataRepresentationId_t id = DDS::XCDR_DATA_REPRESENTATION;
```

File: dds/DCPS/TopicImpl.cpp [hunk 10]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
+    if(qos_ids.length() != 0){
+      id = qos_ids[0];
+    }
+    bool found = false;
+    for (int j = 0; j < type_allowed_reprs.length(); j++){
```
Comment (mitza-oci) [inline dds/DCPS/TopicImpl.cpp:237]:
```suggestion
    for (int j = 0; j < type_allowed_reprs.length(); ++j){
```

File: dds/DCPS/DCPS_Utils.cpp [hunk 11]
```diff
@@ -328,7 +328,7 @@ compatibleQOS(const DDS::DataWriterQos * writerQos,
         }
     }
     }
-    if (reader_count == 0){
+    if (reader_count == 0 {
```
Comment (sonndinh) [inline dds/DCPS/DCPS_Utils.cpp:331]: Missing a `)` here.

File: dds/DCPS/TopicImpl.cpp [hunk 12]
```diff
@@ -256,9 +256,9 @@ bool TopicImpl::check_data_representation(const DDS::DataRepresentationIdSeq& qo
...
   else{
     bool found = false;
-    for (int i = 0; i < qos_ids.length(); i++) {
-      for (int j = 0; j < type_allowed_reprs.length(); j++) {
-        if(qos_ids[i] == type_allowed_reprs[j]) {
+    for (int i = 0; i < qos_ids.length(); ++i) {
+      for (int j = 0; j < type_allowed_reprs.length(); ++j) {
+        if (qos_ids[i] == type_allowed_reprs[j]) {
         found = true;
         break;
```
Comment (sonndinh) [inline dds/DCPS/TopicImpl.cpp]: Missing indentation inside the `if` statement.

File: dds/DCPS/DCPS_Utils.cpp [hunk 13]
```diff
@@ -328,7 +328,7 @@ compatibleQOS(const DDS::DataWriterQos * writerQos,
         }
     }
```
Comment (sonndinh) [inline dds/DCPS/DCPS_Utils.cpp:329]: Looks like the `}` is out of alignment with the `for` loop.

File: dds/DCPS/DCPS_Utils.cpp [hunk 14]
```diff
@@ -315,6 +315,28 @@ compatibleQOS(const DDS::DataWriterQos * writerQos,
...
+    }
+    }
+    if (reader_count == 0 {
+      for (CORBA::ULong wi = 0; !found && wi < writer_count; ++wi) {
+        if (writerQos->representation.value[wi] ==
+            DDS::XCDR_DATA_REPRESENTATION) {
+          found = true;
+          break;
+        }
+    }
```
Comment (sonndinh) [inline dds/DCPS/DCPS_Utils.cpp:338]: Closing `}` of the `for` loop is out of alignment.

File: dds/DCPS/TopicImpl.cpp [hunk 15]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
+    for (int j = 0; j < type_allowed_reprs.length(); ++j) {
+      if (DDS::XCDR_DATA_REPRESENTATION == type_allowed_reprs[j]) {
+        found = true;
+        break;
+      }
+    }
+    return found;
+  }
+  // if non 0 length data reader compare both lists for a compatible QoS
+  else{
```
Comment (sonndinh) [inline dds/DCPS/TopicImpl.cpp:257]:
```suggestion
  else {
```

File: dds/DCPS/TopicImpl.cpp [hunk 16]
```diff
@@ -235,10 +217,56 @@ TopicImpl::inconsistent_topic(int count)
...
+    bool found = false;
+    for (int i = 0; i < qos_ids.length(); ++i) {
+      for (int j = 0; j < type_allowed_reprs.length(); ++j) {
+        if (qos_ids[i] == type_allowed_reprs[j]) {
+        found = true;
+        break;
```
Comment (sonndinh) [inline dds/DCPS/TopicImpl.cpp:263]:
```suggestion
          found = true;
          break;
```

File: dds/DCPS/TopicImpl.cpp (hunk 12 above)
Comment (ClaytonCalabrese) [reply dds/DCPS/TopicImpl.cpp]: It appears as if this is the case because the if conditional is split across 2 lines. I'll just put it back on one line.

**Pull request merged**